### ISSUE TITLE
 'OAuth UID must be unique per provider' 

### ISSUE THREAD

**Issue by infoelliex** [it-projects-llc/odoo-saas-tools#40]
I also receive the "Status Code - 404
Reason - NOT FOUND URL - http://s1.odoo.local:8069/saas_server/new_database?" error when I click the create template database. 

I have followed Ivan's instruction and read all the support docs including the "issue in step6 (https://github.com/yelizariev/odoo-saas-tools/issues/14#issuecomment-134554742)" religiously, but still  cannot solve the problem. 

It was told that we may "....changed the apache config so that it will pass host header and it works." but I have change and try so many settings based on the instruction  (http://httpd.apache.org/docs/2.2/vhosts/examples.html) but still getting the same error: 

I try modify apache's /site-available/000-default.conf by adding

ProxyPreserveHost On
ProxyPass / http://0.0.0.0:8069/
ProxyPassReverse / http://0.0.0.0:8069/
ServerName localhost

I replace the server name with ip or real domain name, also the 0.0.0.0 with ip address or real domain name, adding different port 8069, 8070 etc, or add /saas_server/new_database to the list and etc but none of them working. I still getting the same 404 not found. 

Is it possible to add the web server setting into this instruction? Thank you. 

Error message when try to save the template database:

  File "/opt/odoo/addons/saas_portal/models/saas_portal.py", line 292, in create_template
    raise Warning(msg)
Warning: Status Code - 404
Reason - NOT FOUND
URL - http://s1.odoo.local:8069/saas_server/new_database?access_token=FiR3IICuQe93NFC8OdjtTluHFD1FhM&expires_in=3600&token_type=Bearer&state=%7B%22lang%22%3A+%22en_US%22%2C+%22tz%22%3A+false%2C+%22d%22%3A+%22t2.paulfan.net%22%2C+%22demo%22%3A+0%2C+%22is_template_db%22%3A+1%2C+%22addons%22%3A+%5B%5D%7D&redirect_uri=http%3A%2F%2Fs1.paulfan.net%3A8069%2Fsaas_server%2Fnew_database&response_type=token&client_id=e593331c-6167-11e5-8546-04016e5a6f01&scope=userinfo+force_login+trial+skiptheuse

I also skipped the step 6 through 9 and go directly went to the "page/start" page on the URL, then doing that, I got the 500 Internal Server error: 

500: Internal Server Error
Error
Error message:
There is no plan configured
Traceback

I think this is because the template database cannot be saved on the previous steps / or something else. I appreciate if you could help on this. Thank you very much in advance : )

**Comment by yelizariev** [comment]
if you use port in url (...:8069), then the issue is not related to apache.

Check carefully step 1 of the instruction, expecially about /etc/hosts

**Comment by infoelliex** [comment]
Thank you for fast response. I redo the Step 1, but I still get the same 404 error. I changed the entire odoo from my Linux virtual machine to the Ubuntu at Digital Ocean, and I still get the same 404 error. 

I have /etc/hosts and nginx configured exactly on your instruction. I also experimented different type of settings and rebooted the VPS multiple times, clear browser caches to make sure nothing mistakes made. But I still see the 404 error either trying to create the template database or logging in through the saas portal.

**Comment by yelizariev** [comment]
Where did you configure /etc/hosts ? It should be on machine where you have odoo installed.

Send to us output of following commands on server

```
netstat -lnput

ping -c3 - s1.odoo.local
```

**Comment by infoelliex** [comment]
I reload the Odoo (Ubuntu 14.4) and reconfigured everything and now instead of 404 error , I got this error: 

( I have copied pasted the "netstat and ping" output at the end.)

Here is the odoo error message when trying to create template DB (POS + ECommerce). 

File "/usr/lib/python2.7/dist-packages/requests/adapters.py", line 378, in send
    raise ConnectionError(e)
ConnectionError: HTTPConnectionPool(host='s1.odoo.local', port=8069): Max retries exceeded with url: /saas_server/new_database?access_token=uGTYRRT8ZctLI2nIm3vd3Nwjik3NQs&expires_in=3600&token_type=Bearer&state=%7B%22lang%22%3A+%22en_US%22%2C+%22tz%22%3A+false%2C+%22d%22%3A+%22+t1.odoo.local%22%2C+%22demo%22%3A+0%2C+%22is_template_db%22%3A+1%2C+%22addons%22%3A+%5B%5D%7D&redirect_uri=http%3A%2F%2Fs1.odoo.local%3A8069%2Fsaas_server%2Fnew_database&response_type=token&client_id=125a3118-64d1-11e5-84b1-0800278b5196&scope=userinfo+force_login+trial+skiptheuse (Caused by <class 'socket.gaierror'>: [Errno -2] Name or service not known)

Here is the netstat and ping output: 

osboxes@osboxes:/odoo/odoo-server$ netstat -lnput

(Not all processes could be identified, non-owned process info
 will not be shown, you would have to be root to see it all.)
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 127.0.1.1:53            0.0.0.0:\*               LISTEN      -  
tcp        0      0 127.0.0.1:631           0.0.0.0:\*               LISTEN      -  
tcp        0      0 127.0.0.1:5432          0.0.0.0:\*               LISTEN      -  
tcp        0      0 0.0.0.0:8069            0.0.0.0:\*               LISTEN      -  
tcp        0      0 0.0.0.0:80              0.0.0.0:\*               LISTEN      -  
tcp6       0      0 ::1:631                 :::\*                    LISTEN      -  
tcp6       0      0 :::80                   :::\*                    LISTEN      -  
udp        0      0 0.0.0.0:631             0.0.0.0:\*                           -  
udp        0      0 0.0.0.0:54327           0.0.0.0:\*                           -  
udp        0      0 0.0.0.0:5353            0.0.0.0:\*                           -  
udp        0      0 0.0.0.0:4094            0.0.0.0:\*                           -  
udp        0      0 127.0.1.1:53            0.0.0.0:\*                           -  
udp        0      0 0.0.0.0:68              0.0.0.0:\*                           -  
udp6       0      0 :::41794                :::\*                                -  
udp6       0      0 :::5353                 :::\*                                -  
udp6       0      0 :::7420                 :::\*                                -               

osboxes@osboxes:/odoo/odoo-server$ ping -c3 - s1.odoo.local
ping: unknown host -
osboxes@osboxes:/odoo/odoo-server$ 

This is my etc/hosts: 

127.0.0.1:8069  localhost
127.0.1.1   osboxes

# The following lines are desirable for IPv6 capable hosts

::1     ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters

127.0.0.1:8069 odoo.local # portal
127.0.0.1:8069 s1.odoo.local # server
127.0.0.1:8069 t1.odoo.local # template
127.0.0.1:8069 t2.odoo.local # template
127.0.0.1:8069 client-x.odoo.local
127.0.0.1:8069 client-y.odoo.local
127.0.0.1:8069 client-z.odoo.local

This is the nginx settings: 

# You may add here your

# server {

# ...

# }

# statements for each of your virtual hosts to this file

##

# You should look at the following URL's in order to grasp a solid understanding

# of Nginx configuration files in order to fully unleash the power of Nginx.

# http://wiki.nginx.org/Pitfalls

# http://wiki.nginx.org/QuickStart

# http://wiki.nginx.org/Configuration

#

# Generally, you will want to move this file somewhere, and start with a clean

# file but keep this around for reference. Or just disable in sites-enabled.

#

# Please see /usr/share/doc/nginx-doc/examples/ for more detailed examples.

##

server {
    listen 80 default_server;
    listen [::]:80 default_server ipv6only=on;

```
root /usr/share/nginx/html;
index index.html index.htm;

# Make site accessible from http://localhost/
server_name localhost;

location / {
    # First attempt to serve request as file, then
    # as directory, then fall back to displaying a 404.
    try_files $uri $uri/ =404;
    # Uncomment to enable naxsi on this location
    # include /etc/nginx/naxsi.rules
}

# Only for nginx-naxsi used with nginx-naxsi-ui : process denied requests
#location /RequestDenied {
#   proxy_pass http://127.0.0.1:8080;    
#}

#error_page 404 /404.html;

# redirect server error pages to the static page /50x.html
#
#error_page 500 502 503 504 /50x.html;
#location = /50x.html {
#   root /usr/share/nginx/html;
#}

# pass the PHP scripts to FastCGI server listening on 127.0.0.1:9000
#
#location ~ \.php$ {
#   fastcgi_split_path_info ^(.+\.php)(/.+)$;
#   # NOTE: You should have "cgi.fix_pathinfo = 0;" in php.ini
#
#   # With php5-cgi alone:
#   fastcgi_pass 127.0.0.1:9000;
#   # With php5-fpm:
#   fastcgi_pass unix:/var/run/php5-fpm.sock;
#   fastcgi_index index.php;
#   include fastcgi_params;
#}

# deny access to .htaccess files, if Apache's document root
# concurs with nginx's one
#
#location ~ /\.ht {
#   deny all;
#}
```

}

# another virtual host using mix of IP-, name-, and port-based configuration

#
#server {
    listen 8069;
    listen somename:8069;

# server_name somename alias another.alias;

# root html;

# index index.html index.htm;

#

# location / {

# try_files $uri $uri/ =404;

# }

```
location / {
proxy_pass http://odoo.local:8069;
proxy_next_upstream error timeout invalid_header http_500 http$
proxy_redirect off;

    proxy_set_header Host $host;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_fo$
    proxy_set_header X-Forwarded-Proto $scheme;
```

}

#}

# HTTPS server

#
#server {

# listen 443;

# server_name localhost;

#

# root html;

# index index.html index.htm;

#

# ssl on;

# ssl_certificate cert.pem;

# ssl_certificate_key cert.key;

#

# ssl_session_timeout 5m;

#

# ssl_protocols SSLv3 TLSv1 TLSv1.1 TLSv1.2;

# ssl_ciphers "HIGH:!aNULL:!MD5 or HIGH:!aNULL:!MD5:!3DES";

# ssl_prefer_server_ciphers on;

#

# location / {

# try_files $uri $uri/ =404;

# }

#}

**Comment by yelizariev** [comment]
You have a real mess.

Don't specify port at /etc/hosts
Don't use "listen 8069" at nginx\apache. Use "listen 80" or don't use nginx\apache at all.

**Comment by infoelliex** [comment]
Thank you very much. 

And actually, I did not set any 8069 port in the /etc/hosts or nginx at the first place. It is because of first 404 error, then I set the 8069 port in the hosts and nginx files. 

I change the setting back by removing the 8069 port in etc/hosts and in the nginx configs, then the original 404 not found error returns back. 

The s1.odoo.local is pinged okay in the terminal. I can type s1.odoo.local:8069 on the firefox and it show the default database selection screen. 

![image](https://cloud.githubusercontent.com/assets/14144782/10123889/945b8ea4-6513-11e5-85a9-fb75e6c7a319.png)

 But if I try to save the template database from here: 

![image](https://cloud.githubusercontent.com/assets/14144782/10123895/d9d40e8e-6513-11e5-97ba-6e5015cd223e.png)

I will get the 404 not found error again. It seems like the s1.odoo.local is okay but the string after that (.../saas_server/new_database/....) was not found by the odoo. 

![image](https://cloud.githubusercontent.com/assets/14144782/10123899/121e5db2-6514-11e5-9935-ab1305552585.png)

I have also checked all previous steps and make sure I don't make any other configuration mistakes. I make sure all saas_portal_.. and others are installed.  I test few hours and still cannot get pass to this template database creation step. 

![image](https://cloud.githubusercontent.com/assets/14144782/10123947/488bcda2-6515-11e5-945b-53c61ff5bf8b.png)

I will try to set up the server to only listen to port 80 (instead of 8069) and see if that make any difference.

**Comment by infoelliex** [comment]
<1>. After a full day of testing and twisting, I still cannot get it to work. I saw the same 404 error when click the create template DB red bottom. I even add the server1.com to the /etc/hosts and update the port to 8069 in the SasS/SaaS/Server and click create template DB red bottom and it also get the 404 error as well.

2015-09-28 03:15:47,004 1274 ERROR odoo.local openerp.http: Exception during JSON request handling.
Traceback (most recent call last):
  File "/odoo/odoo-server/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/odoo/odoo-server/openerp/http.py", line 574, in dispatch
    result = self._call_function(*_self.params)
  File "/odoo/odoo-server/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, *_kwargs)
  File "/odoo/odoo-server/openerp/service/model.py", line 113, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/http.py", line 307, in checked_call
    return self.endpoint(_a, *_kw)
  File "/odoo/odoo-server/openerp/http.py", line 803, in __call__
    return self.method(_args, *_kw)
  File "/odoo/odoo-server/openerp/http.py", line 403, in response_wrap
    response = f(_args, *_kw)
  File "/odoo/odoo-server/addons/web/controllers/main.py", line 952, in call_button
    action = self._call_kw(model, method, args, {})
  File "/odoo/odoo-server/addons/web/controllers/main.py", line 940, in _call_kw
    return checked_call(request.db, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/service/model.py", line 113, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/odoo/odoo-server/addons/web/controllers/main.py", line 939, in checked_call
    return getattr(request.registry.get(model), method)(request.cr, request.uid, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/api.py", line 241, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/api.py", line 363, in old_api
    result = method(recs, _args, *_kwargs)
  File "/odoo/custom/addons/saas_portal/models/saas_portal.py", line 292, in create_template
    raise Warning(msg)
Warning: Status Code - 404
Reason - NOT FOUND
URL - http://s1.odoo.local:8069/saas_server/new_database?access_token=l6oowMY2A3JZIL0gsRUoNckIkSIwqR&expires_in=3600&token_type=Bearer&state=%7B%22lang%22%3A+%22en_US%22%2C+%22tz%22%3A+false%2C+%22d%22%3A+%22+t1.odoo.local%22%2C+%22demo%22%3A+0%2C+%22is_template_db%22%3A+1%2C+%22addons%22%3A+%5B%5D%7D&redirect_uri=http%3A%2F%2Fs1.odoo.local%3A8069%2Fsaas_server%2Fnew_database&response_type=token&client_id=125a3118-64d1-11e5-84b1-0800278b5196&scope=userinfo+force_login+trial+skiptheuse

I then un-install nginx, reboot the VM and redo the create template DB procedure, but I still get the same 404 error as follow:

If I click log in via SaaS portal, it will show this URL on the browser: http://odoo.local:8069/oauth2/auth?state=%7B%22p%22%3A+4%2C+%22r%22%3A+%22http%253A%252F%252Fs1.odoo.local%253A8069%252Fweb%22%2C+%22d%22%3A+%22s1.odoo.local%22%7D&redirect_uri=http%3A%2F%2Fs1.odoo.local%3A8069%2Fauth_oauth%2Fsignin&response_type=token&client_id=5b347dc8-64cf-11e5-83d5-0800278b5196&debug=False&scope=userinfo

and on the screen it will show the 

"Not Found

The requested URL was not found on the server. If you entered the URL manually please check your spelling and try again."

and the logs show the same 404 not found message:

2015-09-28 03:27:17,644 3081 INFO None werkzeug: 127.0.0.1 - - [28/Sep/2015 03:27:17] "GET /oauth2/auth?state=%7B%22p%22%3A+4%2C+%22r%22%3A+%22http%253A%252F%252Fs1.odoo.local%253A8069%252Fweb%22%2C+%22d%22%3A+%22s1.odoo.local%22%7D&redirect_uri=http%3A%2F%2Fs1.odoo.local%3A8069%2Fauth_oauth%2Fsignin&response_type=token&client_id=5b347dc8-64cf-11e5-83d5-0800278b5196&debug=False&scope=userinfo HTTP/1.1" 404 

I am sure the /etc/hosts is correct: 

127.0.0.1       localhost
127.0.1.1       osboxes

::1     ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters

127.0.0.1 odoo.local # portal
127.0.0.1 s1.odoo.local # server
127.0.0.1 t1.odoo.local # template
127.0.0.1 t2.odoo.local # template
127.0.0.1 client-x.odoo.local
127.0.0.1 client-y.odoo.local
127.0.0.1 client-z.odoo.local
127.0.0.1 server1.com          

<2>. I know the etc/hosts is correct because when I type s1.odoo.local or t1.odoo.local on the browser URL directly, I can see the odoo's database selection screen. 

For example, if I type t3.odoo.local (since it is not in the etc/hosts list), the odoo will says the server not found. so the etc/hosts is correct.

<3>. I notice that when I type "http://s1.odoo.local:8069/web/database/selector" it will show the odoo select database screen. 

Bit if I copy the link to the browser directly: "http://s1.odoo.local:8069/saas_server/new_database?access_token=qnN6E751iX6R6wZNtBWnItKuP9mg4j&expires_in=3600&token_type=Bearer&state={%22lang%22%3A+%22en_US%22%2C+%22tz%22%3A+false%2C+%22d%22%3A+%22+t1.odoo.local%22%2C+%22demo%22%3A+0%2C+%22is_template_db%22%3A+1%2C+%22addons%22%3A+[]}&redirect_uri=http%3A%2F%2Fs1.odoo.local%3A8069%2Fsaas_server%2Fnew_database&response_type=token&client_id=125a3118-64d1-11e5-84b1-0800278b5196&scope=userinfo+force_login+trial+skiptheuse" 

it will show the 500 Internal Server error as follows: 

Traceback (most recent call last):
  File "/odoo/odoo-server/addons/website/models/ir_http.py", line 199, in _handle_exception
    response = super(ir_http, self)._handle_exception(exception)
  File "/odoo/odoo-server/openerp/addons/base/ir/ir_http.py", line 145, in _handle_exception
    return request._handle_exception(exception)
  File "/odoo/odoo-server/openerp/http.py", line 666, in _handle_exception
    return super(HttpRequest, self)._handle_exception(exception)
  File "/odoo/odoo-server/openerp/addons/base/ir/ir_http.py", line 171, in _dispatch
    result = request.dispatch()
  File "/odoo/odoo-server/openerp/http.py", line 684, in dispatch
    r = self._call_function(*_self.params)
  File "/odoo/odoo-server/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, *_kwargs)
  File "/odoo/odoo-server/openerp/service/model.py", line 113, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/http.py", line 307, in checked_call
    return self.endpoint(_a, *_kw)
  File "/odoo/odoo-server/openerp/http.py", line 803, in __call__
    return self.method(_args, *_kw)
  File "/odoo/odoo-server/openerp/http.py", line 403, in response_wrap
    response = f(_args, *_kw)
  File "/odoo/odoo-server/addons/auth_oauth/controllers/main.py", line 42, in wrapper
    return func(self, _a, *_kw)
  File "/odoo/custom/addons/saas_server/controllers/main.py", line 53, in new_database
    client.create_database(template_db, demo, lang)
  File "/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/api.py", line 407, in new_api
    result = [method(rec, _args, *_kwargs) for rec in self]
  File "/odoo/custom/addons/saas_server/models/saas_server.py", line 49, in create_database
    openerp.service.db.exp_create_database(new_db, demo, lang, user_password=password)
  File "/odoo/odoo-server/openerp/service/db.py", line 90, in exp_create_database
    _create_empty_database(db_name)
  File "/odoo/odoo-server/openerp/service/db.py", line 82, in _create_empty_database
    raise DatabaseExists("database %r already exists!" % (name,))
DatabaseExists: database u' t1.odoo.local' already exists!

2015-09-28 03:13:46,331 1274 INFO s1.odoo.local werkzeug: 127.0.0.1 - - [28/Sep/2015 03:13:46] "GET /saas_server/new_database?access_token=LsNMnIHPMNjTE57UcokcgTDdn2VLx3&expires_in=3600&token_type=Bearer&state={%22lang%22%3A+%22en_US%22%2C+%22tz%22%3A+false%2C+%22d%22%3A+%22+t1.odoo.local%22%2C+%22demo%22%3A+0%2C+%22is_template_db%22%3A+1%2C+%22addons%22%3A+[]}&redirect_uri=http%3A%2F%2Fs1.odoo.local%3A8069%2Fsaas_server%2Fnew_database&response_type=token&client_id=125a3118-64d1-11e5-84b1-0800278b5196&scope=userinfo+force_login+trial+skiptheuse HTTP/1.1" 500 -

And if I only I type "http://s1.odoo.local/saas_server/new_database?" on the browser, it will redirect me to the "http://s1.odoo.local:8069/" and no error messages display. 

So I believe the s1.odoo.local is good. Even the s1.odoo.local/saas_server/new_database? is good. 

However, the long string after that (/new_database?...), for some reason, is not being accepted by the odoo.

<4>. Since I have remove nginx and no apache was installed in this new image. I assume that I don't need to worry about these setting from step 1, is that correct?
   proxy_set_header Host $host;
   proxy_set_header X-Real-IP $remote_addr;
   proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
   proxy_set_header X-Forwarded-Proto $scheme;

What else I can try and test? Do you have any suggestion? I appreciate your help on this matter. 

Thank you again.

p.s. I have followed the conversation in the "issue in step 6: Create plan #14" to make sure all the configuration are correct. I just don't know if I need to reinstall the apache again since it was saying that the apache config fixed the problem (but you say the nginx or apache are not needed for this project to run, so I didn't install the apache in the new ubuntu image I create yesterday).

p.p.s. I will get the same 404 error if I try to sync server: http://s1.odoo.local:8069/saas_server/sync_server?..... as follows:

Warning: Status Code - 404
Reason - NOT FOUND
URL - http://s1.odoo.local:8069/saas_server/sync_server?access_token=r2RYuRY93vBfhTAER0Zhayvgahrnm7&expires_in=3600&token_type=Bearer&state=%7B%22d%22%3A+%22s1.odoo.local%22%2C+%22client_id%22%3A+%225b347dc8-64cf-11e5-83d5-0800278b5196%22%7D&redirect_uri=http%3A%2F%2Fs1.odoo.local%3A8069%2Fweb&response_type=token&client_id=5b347dc8-64cf-11e5-83d5-0800278b5196&scope=userinfo+force_login+trial+skiptheuse

p.p.p.s.

When I use the sudo service odoo-server start, everything is okay. 
but if I use the ./odoo.py --db-filter '^%h$' or .... ^%h$ or %h to start the odoo server, I will also get 500 internal error (I use osbox.org ubuntu 14 image): 

```
connection_factory=connection_factory, async=async)
```

OperationalError: FATAL:  role "osboxes" does not exist
2015-09-28 03:53:54,048 3856 ERROR None openerp.sql_db: Connection to the database failed
Traceback (most recent call last):
  File "/odoo/odoo-server/openerp/sql_db.py", line 526, in borrow
    result = psycopg2.connect(dsn=dsn, connection_factory=PsycoConnection)
  File "/usr/lib/python2.7/dist-packages/psycopg2/**init**.py", line 179, in connect

I try to add dbfilter = '^%h$' or try = %h. Either way, it will make me stuck in the "http://127.0.0.1:8069/web/database/manager#action=database_manager" screen. I can create more database in that screen, but I will never be able to come out of that screen unless I remove the dbfilter = '^%h$' setting in the odoo-server.conf.

**Comment by yelizariev** [comment]
Try to follow instruction, i.e. don't specify 8069 port in SaaS Server and use nginx

**Comment by infoelliex** [comment]
I manage to get some parts of it work. The create template DB work now but I have to use a work around. I will explain later.

The main issue is now that I got another error when trying to create client database from template DB. I got 500 internal error on OAuth UID not unique (however, I can still see the client database in the servers under the s1.odoo.local, but the data base is not created). I try couple times and all get 500 error. I can see lots of same client-x.odoo.local under the s1.odoo.local. When I try to delete one of them, the error say the client data base does not exist. So the client data base never created due to the 500 error as follow:

```
return checked_call(self.db, *args, **kwargs)
```

  File "/odoo/odoo-server/openerp/service/model.py", line 129, in wrapper
    raise openerp.osv.orm.except_orm(_('Constraint Error'), tr(registry._sql_error[key], 'sql_constraint') or inst[0])
except_orm: ('Constraint Error', 'OAuth UID must be unique per provider')

2015-09-28 04:42:42,402 5057 INFO s1.odoo.local werkzeug: 127.0.0.1 - - [28/Sep/2015 04:42:42] "GET /saas_server/new_database?access_token=0VOviR7BzOYvUJMnM6KhaPey7TZh0Y&token_type=Bearer&state=%7B%22db_template%22%3A+%22t2.odoo.local%22%2C+%22r%22%3A+%22http%3A%2F%2Fc2.odoo.local%3A8069%2Fweb%22%2C+%22e%22%3A+false%2C+%22d%22%3A+%22c2.odoo.local%22%7D&expires_in=3600&scope=userinfo+force_login+trial+skiptheuse HTTP/1.1" 500 -

**Comment by infoelliex** [comment]
Talk about the create template DB work around: 

I was always restarting the server using the $sudo service odoo-server restart 
and then I also use $./odoo.py --db-filter ^%h$ (although I got odoo already running but I ignore it)

Then I do this as you teach in the step 1 "sed -i 's/matches="[^"]*"//g' addons/web/static/src/xml/base.xml" all good w/o any error. 

Since the 404 error occurs, I try another way: 

I add ^%h$ to the odoo-server.conf as follows:

dbfilter = ^%h$ (I also try dbfilter = '^%h$' and dbfilter = %h) 
all seems are the same.

Then I restart the server, but when I go to the browser, I only can see the manage database screen. I then create new database but no matter how I try, the odoo will keep me in the same database screen. Even I type the database selector, I still cannot get out. 

Then I copy one of the longer name longer URL that I open in another tab (http://odoo.local:8069/web#page=0&limit=80&view_type=list&model=saas_portal.plan&menu_id=123&action=121) and that let me out of the database management screen. 

If I remove the dbfilter = ^%h$ in the odoo-server.conf, I will also able to get out of the database manage screen but then I will get the 400 not found error again when create template DB. 

So I guess, this is a workaround, but not a real solution to the problem. In this case, I also only can login again with email (admin) and pwd. The database selection will be gone from the log in screen.

But at least I can get the create template DB work this way. Is there better way to configure dbfilter = ? Thank you for helping.

The current status is that I can create template DB but not the client database because of the 500 error indicating that the OAuth UID must be unique per provider.

I will also follow your instruction, i.e." don't specify 8069 port in SaaS Server and use nginx" >>> do you mean to use the nginx proxy to forward 8069 to 80 so I can use odoo.local instead of odoo.local:8069. I did that too before, but I will do it again to see if there would make any difference at this point. Thank you.

**Comment by yelizariev** [comment]
Try to start fresh databases (main, server, template) and repeat the process. I if the issue  'OAuth UID must be unique per provider'  happens again, I would reopen the issue

**Comment by infoelliex** [comment]
1. I added the server2.com, s1.server2.com, t1.. and client-x ... in the /etc/hosts (reboot VBox okay)
2. I create the main database server2.com (ok) then I install the saas_portal_*

(>>> problem) 3. when I create the server database s1.server2.com. It was ok too. But the odoo redirect me to a login page that has no database selection again? So I can only login to the server2.com database but not the s1.server2.com. 
1. The only way to go around with that is to manually type "http://s1.server2.com:8069" on the browser. 
2. Then I install the saas_server (ok)
   (>>> problem) 6. Again when follow step 4, when log out the administrator, I cannot see the database selection, so I have to clear the browser cache and type server2.com:8069 to login to the main database. 
3. In the step 6 when I try to [Create Template DB], i receive this error: (don't know why this happen? I was doing odoo.local create template DB just fine. I didn't change any odoo-server.conf settings but only add server2.com, s1.server2.com and etc to the etc/hosts).

2015-09-28 06:46:11,162 1460 INFO s1.server2.com werkzeug: 127.0.0.1 - - [28/Sep/2015 06:46:11] "GET /saas_server/new_database?access_token=nfLCJctBMG1tZ5YPsfEVCP1EtfrBVj&expires_in=3600&token_type=Bearer&state=%7B%22lang%22%3A+%22en_US%22%2C+%22tz%22%3A+false%2C+%22d%22%3A+%22t1.server2.com%22%2C+%22demo%22%3A+0%2C+%22is_template_db%22%3A+1%2C+%22addons%22%3A+%5B%5D%7D&redirect_uri=http%3A%2F%2Fs1.server2.com%3A8069%2Fsaas_server%2Fnew_database&response_type=token&client_id=57f8e388-65ab-11e5-ba64-0800278b5196&scope=userinfo+force_login+trial+skiptheuse HTTP/1.1" 500 -
2015-09-28 06:46:11,165 1460 ERROR server2.com openerp.http: Exception during JSON request handling.
Traceback (most recent call last):
  File "/odoo/odoo-server/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/odoo/odoo-server/openerp/http.py", line 574, in dispatch
    result = self._call_function(*_self.params)
  File "/odoo/odoo-server/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, *_kwargs)
  File "/odoo/odoo-server/openerp/service/model.py", line 113, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/http.py", line 307, in checked_call
    return self.endpoint(_a, *_kw)
  File "/odoo/odoo-server/openerp/http.py", line 803, in __call__
    return self.method(_args, *_kw)
  File "/odoo/odoo-server/openerp/http.py", line 403, in response_wrap
    response = f(_args, *_kw)
  File "/odoo/odoo-server/addons/web/controllers/main.py", line 952, in call_button
    action = self._call_kw(model, method, args, {})
  File "/odoo/odoo-server/addons/web/controllers/main.py", line 940, in _call_kw
    return checked_call(request.db, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/service/model.py", line 113, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/odoo/odoo-server/addons/web/controllers/main.py", line 939, in checked_call
    return getattr(request.registry.get(model), method)(request.cr, request.uid, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/api.py", line 241, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/api.py", line 363, in old_api
    result = method(recs, _args, *_kwargs)
  File "/odoo/custom/addons/saas_portal/models/saas_portal.py", line 292, in create_template
    raise Warning(msg)
Warning: Status Code - 500
Reason - INTERNAL SERVER ERROR
URL - http://s1.server2.com:8069/saas_server/new_database?access_token=nfLCJctBMG1tZ5YPsfEVCP1EtfrBVj&expires_in=3600&token_type=Bearer&state=%7B%22lang%22%3A+%22en_US%22%2C+%22tz%22%3A+false%2C+%22d%22%3A+%22t1.server2.com%22%2C+%22demo%22%3A+0%2C+%22is_template_db%22%3A+1%2C+%22addons%22%3A+%5B%5D%7D&redirect_uri=http%3A%2F%2Fs1.server2.com%3A8069%2Fsaas_server%2Fnew_database&response_type=token&client_id=57f8e388-65ab-11e5-ba64-0800278b5196&scope=userinfo+force_login+trial+skiptheuse

2015-09-28 06:46:11,169 1460 INFO server2.com werkzeug: 127.0.0.1 - - [28/Sep/2015 06:46:11] "POST /web/dataset/call_button HTTP/1.1" 200 -
2015-09-28 06:46:11,214 1460 ERROR s1.server2.com werkzeug: Error on request:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/werkzeug/serving.py", line 177, in run_wsgi
    execute(self.server.app)
  File "/usr/lib/python2.7/dist-packages/werkzeug/serving.py", line 165, in execute
    application_iter = app(environ, start_response)
  File "/odoo/odoo-server/openerp/service/server.py", line 290, in app
    return self.app(e, s)
  File "/odoo/odoo-server/openerp/service/wsgi_server.py", line 216, in application
    return application_unproxied(environ, start_response)
  File "/odoo/odoo-server/openerp/service/wsgi_server.py", line 202, in application_unproxied
    result = handler(environ, start_response)
  File "/odoo/odoo-server/openerp/http.py", line 1290, in **call**
    return self.dispatch(environ, start_response)
  File "/odoo/odoo-server/openerp/http.py", line 1264, in **call**
    return self.app(environ, start_wrapped)
  File "/usr/lib/python2.7/dist-packages/werkzeug/wsgi.py", line 579, in **call**
    return self.app(environ, start_response)
  File "/odoo/odoo-server/openerp/http.py", line 1437, in dispatch
    result = ir_http._dispatch()
  File "/odoo/odoo-server/addons/website/models/ir_http.py", line 148, in _dispatch
    resp = super(ir_http, self)._dispatch()
  File "/odoo/odoo-server/openerp/addons/base/ir/ir_http.py", line 175, in _dispatch
    return self._handle_exception(e)
  File "/odoo/odoo-server/addons/website/models/ir_http.py", line 248, in _handle_exception
    html = request.website._render('website.http_error', values)
  File "/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/api.py", line 556, in new_api
    result = method(self._model, cr, uid, self.ids, _args, *_kwargs)
  File "/odoo/odoo-server/addons/website/models/website.py", line 288, in _render
    return self.pool['ir.ui.view'].render(cr, uid, template, values=values, context=context)
  File "/odoo/odoo-server/openerp/api.py", line 241, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/odoo/odoo-server/addons/website/models/ir_ui_view.py", line 166, in render
    view_obj = request.website.get_template(id_or_xml_id)
  File "/odoo/odoo-server/openerp/api.py", line 239, in wrapper
    return new_api(self, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/api.py", line 556, in new_api
    result = method(self._model, cr, uid, self.ids, _args, *_kwargs)
  File "/odoo/odoo-server/addons/website/models/website.py", line 283, in get_template
    model, view_id = request.registry["ir.model.data"].get_object_reference(cr, uid, module, xmlid)
  File "/odoo/odoo-server/openerp/api.py", line 241, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/addons/base/ir/ir_model.py", line 968, in get_object_reference
    return self.xmlid_lookup(cr, uid, "%s.%s" % (module, xml_id))[1:3]
  File "/odoo/odoo-server/openerp/api.py", line 241, in wrapper
    return old_api(self, _args, *_kwargs)
  File "<string>", line 2, in xmlid_lookup

  File "/odoo/odoo-server/openerp/tools/cache.py", line 74, in lookup
    value = d[key] = self.method(_args, *_kwargs)
  File "/odoo/odoo-server/openerp/addons/base/ir/ir_model.py", line 924, in xmlid_lookup
    ids = self.search(cr, uid, [('module','=',module), ('name','=', name)])
  File "/odoo/odoo-server/openerp/api.py", line 241, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/models.py", line 1645, in search
    return self._search(cr, user, args, offset=offset, limit=limit, order=order, context=context, count=count)
  File "/odoo/odoo-server/openerp/api.py", line 241, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/models.py", line 4682, in _search
    cr.execute(query_str, where_clause_params)
  File "/odoo/odoo-server/openerp/sql_db.py", line 158, in wrapper
    return f(self, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
InternalError: current transaction is aborted, commands ignored until end of transaction block
1. Note: I got similar 500 error when I try to syn the server on the step 5. 

Odoo Server Error

Traceback (most recent call last):
  File "/odoo/odoo-server/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/odoo/odoo-server/openerp/http.py", line 574, in dispatch
    result = self._call_function(*_self.params)
  File "/odoo/odoo-server/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, *_kwargs)
  File "/odoo/odoo-server/openerp/service/model.py", line 113, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/http.py", line 307, in checked_call
    return self.endpoint(_a, *_kw)
  File "/odoo/odoo-server/openerp/http.py", line 803, in __call__
    return self.method(_args, *_kw)
  File "/odoo/odoo-server/openerp/http.py", line 403, in response_wrap
    response = f(_args, *_kw)
  File "/odoo/odoo-server/addons/web/controllers/main.py", line 952, in call_button
    action = self._call_kw(model, method, args, {})
  File "/odoo/odoo-server/addons/web/controllers/main.py", line 940, in _call_kw
    return checked_call(request.db, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/service/model.py", line 113, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/odoo/odoo-server/addons/web/controllers/main.py", line 939, in checked_call
    return getattr(request.registry.get(model), method)(request.cr, request.uid, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/api.py", line 241, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/api.py", line 403, in old_api
    result = new_api(recs, _args, *_kwargs)
  File "/odoo/odoo-server/openerp/api.py", line 407, in new_api
    result = [method(rec, _args, *_kwargs) for rec in self]
  File "/odoo/custom/addons/saas_portal/models/saas_portal.py", line 123, in action_sync_server
    raise Warning(msg)
Warning: Status Code - 500
Reason - INTERNAL SERVER ERROR
URL - http://s1.server2.com:8069/saas_server/sync_server?access_token=vblMH4FZxUgzKjS8ODhjwXGzVfx6Nl&expires_in=3600&token_type=Bearer&state=%7B%22d%22%3A+%22s1.server2.com%22%2C+%22client_id%22%3A+%22+3fcdd276-65a8-11e5-b3b5-0800278b5196%22%7D&redirect_uri=http%3A%2F%2Fs1.server2.com%3A8069%2Fweb&response_type=token&client_id=+3fcdd276-65a8-11e5-b3b5-0800278b5196&scope=userinfo+force_login+trial+skiptheuse
1. This is my odoo-server.conf setting: I believe it should be fine. 

[options]
; This is the password that allows database operations:
admin_passwd = admin
db_host = False
db_port = False
db_user = odoo
db_password = False
addons_path = /usr/lib/python2.7/dist-packages/openerp/addons
logfile = /var/log/odoo/odoo-server.log
addons_path=/odoo/odoo-server/addons,/odoo/custom/addons
xmlrpc_port = 8069
dbfilter = %h 

note: Is above %h okay. I tried ^%h$ or '^%h$' before and seems no difference at all.

/etc/hosts
127.0.0.1 odoo.local # portal
127.0.0.1 s1.odoo.local # server
127.0.0.1 t1.odoo.local # template
127.0.0.1 t2.odoo.local # template
127.0.0.1 client-x.odoo.local
127.0.0.1 client-y.odoo.local
127.0.0.1 client-z.odoo.local
127.0.0.1 server2.com
127.0.0.1 s1.server2.com
127.0.0.1 t1.server2.com
127.0.0.1 client-x.server2.com

(I have reboot the machine after adding server2.com to the etc/hosts)

Do I get anything wrong on the conf or hosts setting? But I didn't touch these files at all.

Since I cannot get the create template DB to work again (as I have 500 internal error) and I do not see the database option drop down menu in the login screen, I think I might need to reload the odoo 8 from scrach and try again. What would you recommend?  Thank you very much again.

**Comment by yelizariev** [comment]
I recommend you to follow instructions and fix issues you faced rather than trying other configuration (like add 8069 to /etc/hosts or to SaaS server)

**Comment by infoelliex** [comment]
I did not try to do other configuration. I only remove all database and redo the exercise from scratch. And everything was working fine until the step #8 which when I click the create client database (red bottom), I would still get the 'OAuth UID must be unique per provider' 500 server error. In case one test is not enough, so I try this in the Virtual Box and in the Digital Ocean. The result are the same.

![image](https://cloud.githubusercontent.com/assets/14144782/10149775/afa04a82-6609-11e5-8103-b495f2289eb3.png)

Traceback (most recent call last):
  File "/odoo/odoo-server/addons/website/models/ir_http.py", line 199, in _handle_exception
    response = super(ir_http, self)._handle_exception(exception)
  File "/odoo/odoo-server/openerp/addons/base/ir/ir_http.py", line 145, in _handle_exception
    return request._handle_exception(exception)
  File "/odoo/odoo-server/openerp/http.py", line 666, in _handle_exception
    return super(HttpRequest, self)._handle_exception(exception)
  File "/odoo/odoo-server/openerp/addons/base/ir/ir_http.py", line 171, in _dispatch
    result = request.dispatch()
  File "/odoo/odoo-server/openerp/http.py", line 684, in dispatch
    r = self._call_function(*_self.params)
  File "/odoo/odoo-server/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, *_kwargs)
  File "/odoo/odoo-server/openerp/service/model.py", line 129, in wrapper
    raise openerp.osv.orm.except_orm(_('Constraint Error'), tr(registry._sql_error[key], 'sql_constraint') or inst[0])
except_orm: ('Constraint Error', 'OAuth UID must be unique per provider')

After seeing that error, I cannot see any client database being created in the SaaS>SaaS>Clients; but I can see the client-x (this name I use) was created in the database. That client database name just won't show in the SaaS>SasS>Clients.

Although I can manually create a Clients database afterward from SaaS>SaaS>Clients, and able to sync server without error, but I won't be able to login as client database or choose the "Count users" and I have to click "in progress" in order to see client database stay on the page.  

But I think the problem rather than on the addons/auth_oauth/res_users.py, it can be on the http.py having problem to handle exception.

File "/odoo/odoo-server/addons/website/models/ir_http.py", line 199, in _handle_exception
    response = super(ir_http, self)._handle_exception(exception)
  File "/odoo/odoo-server/openerp/addons/base/ir/ir_http.py", line 145, in _handle_exception
    return request._handle_exception(exception)
  File "/odoo/odoo-server/openerp/http.py", line 666, in _handle_exception
    return super(HttpRequest, self)._handle_exception(exception)

P.S. 

Regarding the problem of cannot get out of the database/manager page, I try to solve this by doing different configuration also based on the article in the http://netjunky.net/odoo-super-simple-subdomain-filtering-with-cloudflare/. Instead of doing dbfilter = '^%d$' in the odoo-server.conf, I keep the original setting and dbfilter = .\* in odoo-server.conf, and add the --db-filter=^%h$" in the /etc/init.d/odoo-server like this: 

# Additional options that are passed to the Daemon.

#DAEMON_OPTS="-c $CONFIGFILE"
DAEMON_OPTS="-c $CONFIGFILE --db-filter=^%h$"

And make sure my URL showing the same database name as typed in the database manager creation screen.  After that, I can get out of the database creation screen by just typing my domain name on the URL to login to the new database, however, this does not solve any of the client database creation issue mentioned above.

**Comment by yelizariev** [comment]
Are you sure that you use latest saas modules without modifications?

```
git diff
git show HEAD
```

**Comment by saltonmassally** [comment]
@infoelliex please check your servers and ensure that no orphaned clients exists (clients listed in the several but no in the portal) this happens when you manually delete clients using odoo's unlink. Once that is done sync your servers and see if it throws you this error; if it does not proceed with creating your client USING THE WIZARD, not manually.

My hypothesis here is that if you deleted a client manually it doesn't get removed from the server and for some reason the oauth details also remain in the portal (we need to look into this), creating a new client first syncs the server during which an attempt at creating clients on the server but not listed on the portal which might result in the error you see. @yelizariev is in a better position to comment on this

**Comment by yelizariev** [comment]
I got it.
The bug is appeared after this commit https://github.com/yelizariev/odoo-saas-tools/commit/1be2e758da3c4f81cf148190653a67c615c5bebb
I'll think how to fix it.

Thank @infoelliex for reporting

**Comment by yelizariev** [comment]
@infoelliex update source and restart server.

**Comment by infoelliex** [comment]
Hi Ivan, It works!!! Now I see no auth_oauth error message at the client database creation and the client is successfully listed in the SaaS>SaaS>Client page. I can do the exercise from step 1 to 8 without any problem.  I appreciate your help. Thank you very much. 

I think you might also like to know is that the "Create client" window remains opened after the client database had been successfully created (see img below). 

![image](https://cloud.githubusercontent.com/assets/14144782/10171824/0815b760-66aa-11e5-82da-a2c469af299f.png)

I can just manually close the pop up window. But the thing is that if I click the "Create" button again by mistake, a 500 error will show on the screen. 

P.S. Thank you to Salton's advice as well.

**Comment by infoelliex** [comment]
This is a different issue (or a question only). Should I open a new case for this? 

I have a question regarding to the "Log in via Saas Portal" site, see below:

![image](https://cloud.githubusercontent.com/assets/14144782/10181270/0a2328be-66e4-11e5-960f-ed94a8a5497e.png)

After I click the Log in via Saas Portal link (above), I will be redirected to this page (below), is it suppose to happen this way? Thank you very much. 

![image](https://cloud.githubusercontent.com/assets/14144782/10181291/4d29b826-66e4-11e5-9149-fafc605edf42.png)

I will test the rest of the modules you've created and let you know what I found.

**Comment by saltonmassally** [comment]
Yes, it is normal as it authenticates against the portal (Oauth)... If you are already logged into the portal it won't show you the login form however if you are not it will...

**Comment by yelizariev** [comment]
@infoelliex 
you can open new issue for this:

>  I think you might also like to know is that the "Create client" window remains opened after the client database had been successfully created (see img below).

It's known issue, that is not fixed yet

**Comment by Udramani** [comment]
Hello,
I am having problem while i sync the server and this type of error message is displayed.
![screenshot from 2017-10-27 11-56-19](https://user-images.githubusercontent.com/33111998/32090652-02880938-bae0-11e7-9c18-f39693804a4b.png)

![screenshot from 2017-10-27 11-56-45](https://user-images.githubusercontent.com/33111998/32090663-0c362ec4-bae0-11e7-93bc-888fa9f8982a.png)

i am using apache2 server


<VirtualHost *:80>
ServerAdmin webmaster@localhost
ServerName o9.local
ServerAlias *.o9.local
ProxyRequests Off
ProxyPreserveHost On
<Proxy *>
Order allow,deny
Allow from all

ProxyPass / http://127.0.0.1:8097/
ProxyPassReverse / http://127.0.0.1:8097/
ProxyPreserveHost On
ProxyErrorOverride Off
SetEnv proxy-nokeepalive 1
ErrorLog /var/log/apache2/webclient-error.log
CustomLog /var/log/apache2/webclient-access.log combined
</VirtualHost>


can you help me to find out the solution?
